cm31_ntt/src/ntt.rs

```diff
@@ -0,0 +1,1233 @@
+    if is_power_of((n / 4) as u32, 8) {
+        let (precomp_small, precomp_full) = precomp_for_ntt_r8_hybrid_p(n / 4);
+        let precomp_s4 = precomp_s4(n);
+        return Ok(PrecomputedTwiddles {
+            small: precomp_small,
+            full: precomp_full,
+            stride: precomp_s4,
+        });
+    }
+    
+
+    Err("n must be a power of 2, or of the form 8^k * 2 or 8^k * 4".to_string())
+}
+
+/// Performs a radix-8 NTT on the polynomial f.
+/// This is unoptimised as it does not use any precomputed twiddles.
+/// It is also slower than ntt_r8_hybrid_p() as it allocates Vecs at each level of recursion, and doing so at
+/// the lower levels (to handle 8^1 to 8^5) incurs significant memory allocation overhead.
+/// @param f The coefficients of the polynomial to be transformed.
+/// @param w The n-th root of unity, where n is the length of f.
+/// @return The transformed polynomial in evaluation form.
+pub fn ntt_r8_vec(f: &Vec<CF>, w: CF) -> Vec<CF> {
+    let n = f.len();
+    debug_assert!(n >= 8, "n must be at least 8");
+    debug_assert!(is_power_of(n as u32, 8), "n must be a power of 8");
+
+    fn do_ntt(f: &Vec<CF>, w: CF) -> Vec<CF> {
+        let n = f.len();
+        if n == 1 {
+            return f.clone();
+        }
+
+        let m = n / 8;
+        let w_pow_8 = w.pow(8);
+
+        // Partition f into 8 subarrays.
+        let mut a = vec![vec![CF::zero(); m]; 8];
+
+        let mut res = vec![CF::zero(); n];
+        unsafe {
+            for i in 0..m {
+                let i_8 = i * 8;
+                for j in 0..8 {
+                    a[j][i] = *f.get_unchecked(i_8 + j);
+                }
+            }
+
+            // Recurse
+            let mut ntt_a = Vec::with_capacity(8);
+            for i in 0..8 {
+                ntt_a.push(do_ntt(&a[i], w_pow_8));
+            }
+
+            for k in 0..m {
+                // Calculate twiddle factors
+                let wt   = w.pow(k);
+                let wt2  = wt  * wt;
+                let wt3  = wt2 * wt;
+                let wt4  = wt3 * wt;
+                let wt5  = wt4 * wt;
+                let wt6  = wt5 * wt;
+                let wt7  = wt6 * wt;
+
+                let f0 = *ntt_a.get_unchecked(0).get_unchecked(k);
+                let f1 = *ntt_a.get_unchecked(1).get_unchecked(k);
+                let f2 = *ntt_a.get_unchecked(2).get_unchecked(k);
+                let f3 = *ntt_a.get_unchecked(3).get_unchecked(k);
+                let f4 = *ntt_a.get_unchecked(4).get_unchecked(k);
+                let f5 = *ntt_a.get_unchecked(5).get_unchecked(k);
+                let f6 = *ntt_a.get_unchecked(6).get_unchecked(k);
+                let f7 = *ntt_a.get_unchecked(7).get_unchecked(k);
+
+                let butterfly = ntt_block_8(
+                    f0, f1, f2, f3, f4, f5, f6, f7,
+                    wt, wt2, wt3, wt4, wt5, wt6, wt7
+                );
+
+                *res.get_unchecked_mut(k) = butterfly.0;
```

Comment:
k is in range `0..m` where `m = n / 8`, res is `vec![CF::zero(); n];`. If n is smaller than zero, or not divisible by 8 you can run into undefined behaviour here with index overrun.

The asserts at the start only safeguard against it in debug. You need to either replace the debug asserts with regular asserts or make this function fallible.

Reply:
Thanks for spotting this!

Added `assert!(n > 0 && n % 8 == 0); ` in commit aaddb593 to prevent this issue.

Reply:
This relates to https://github.com/worldfnd/ProveKit/pull/19/files#r2072454803

I also noticed that some functions have `debug_asserts` and some have `asserts` in the beginning - let's make a this consistent across all functions

Reply:
Resolved by using `ensure!` consistently.